bancor_research/bancor_simulator/v3/spec/actions.py

```diff
@@ -665,24 +669,19 @@ def external_protection(self):
         """
         a = self.bnt_trading_liquidity
         b = self.tkn_trading_liquidity
+        c = self.tkn_excess
+        e = self.staked_tkn
         n = self.withdrawal_fee
-        T = self.bnt_sent_to_user
         w = self.external_protection_tkn_balance
         x = self.withdraw_value
-        S = self.tkn_sent_to_user
+        y = x * (1 - n)
         if not self.is_trading_enabled:
-            bnt_sent_to_user = Decimal("0")
-            external_protection_compensation = min(w, x * (1 - n) - S)
-        elif T and w:
-            if T * b > w * a:
-                bnt_sent_to_user = (T * b - w * a) / b
-                external_protection_compensation = w
-            else:
-                bnt_sent_to_user = Decimal("0")
-                external_protection_compensation = T * b / a
+            bnt_sent_to_user = max(Decimal("0"), y - w)
+            external_protection_compensation = min(w, y)
         else:
-            bnt_sent_to_user = T
-            external_protection_compensation = Decimal("0")
+            z = y * (e - b - c) / e
+            bnt_sent_to_user = max(Decimal("0"), a * (z - w) / b)
+            external_protection_compensation = min(w, z)
         return bnt_sent_to_user, external_protection_compensation
```

Comment:
you were correct in pointing out the constant 0 value of T passed in - however, once the correct arguments are passed in the remainder of the method I think is preferable how it was rather than updated with your changes, the reason being that it directly aligns with the spec, in terms of the equations and logical tests, whereas the changes youve proposed “might be OK” in the sense of a simplified implementation, but they obfuscate any direct lineage from the spec documentation… so I’d prefer to keep how it was, with the exception of changing the input arguments... I suggest the following changes:

1.) in method `process_withdrawal` change `self.external_protection()` to `self.external_protection(
                    updated_bnt_liquidity,
                    updated_tkn_liquidity,
                    tkn_sent_to_user,
                    bnt_sent_to_user)`

and then keep `external_protection` how it was, but add the new input arguments as follows:

def external_protection(self,
                            bnt_trading_liquidity,
                            tkn_trading_liquidity,
                            tkn_sent_to_user,
                            bnt_sent_to_user):
        """
        This replaces any BNT that would have been received by the user with TKN.
        """
        a = bnt_trading_liquidity
        b = tkn_trading_liquidity
        n = self.withdrawal_fee
        T = bnt_sent_to_user
        w = self.external_protection_tkn_balance
        x = self.withdraw_value
        S = tkn_sent_to_user
        if not self.is_trading_enabled:
            bnt_sent_to_user = Decimal("0")
            external_protection_compensation = min(w, x * (1 - n) - S)
        elif T and w:
            if T * b > w * a:
                bnt_sent_to_user = (T * b - w * a) / b
                external_protection_compensation = w
            else:
                bnt_sent_to_user = Decimal("0")
                external_protection_compensation = T * b / a
        else:
            bnt_sent_to_user = T
            external_protection_compensation = Decimal("0")
        return bnt_sent_to_user, external_protection_compensation